smart-contract-verifier/src/solidity/client.rs

```diff
@@ -0,0 +1,80 @@
+use super::compiler::SolidityCompiler;
+use crate::{compiler::Compilers, middleware::Middleware, verifier::Success};
+use std::sync::Arc;
+
+pub struct ClientBuilder {
+    compilers: Arc<Compilers<SolidityCompiler>>,
+    middleware_stack: Vec<Arc<dyn Middleware<Success>>>,
+}
+
+impl ClientBuilder {
+    /// Convenience method to initialize new client builder.
+    ///
+    /// If you need to keep a reference to the compilers after initialization, use [`new_arc`].
+    ///
+    /// [`new_arc`]: Self::new_arc
+    pub fn new(compilers: Compilers<SolidityCompiler>) -> Self {
+        Self::new_arc(Arc::new(compilers))
+    }
+
+    /// Initialize new client builder. [`new`] is more ergonomic if you don't need the `Arc`.
+    ///
+    /// [`new`]: Self::new
+    pub fn new_arc(compilers: Arc<Compilers<SolidityCompiler>>) -> Self {
+        Self {
+            compilers,
+            middleware_stack: vec![],
+        }
+    }
+
+    /// Convenience method to attach middleware.
+    ///
+    /// If you need to keep a reference to the middleware after attaching, use [`with_arc`].
+    ///
+    /// [`with_arc`]: Self::with_arc
+    pub fn with<M>(self, middleware: M) -> Self
+    where
+        M: Middleware<Success>,
+    {
+        self.with_arc(Arc::new(middleware))
+    }
+
+    /// Add middleware to the chain. [`with`] is more ergonomic if you don't need the `Arc`.
+    ///
+    /// [`with`]: Self::with
+    pub fn with_arc(mut self, middleware: Arc<dyn Middleware<Success>>) -> Self {
+        self.middleware_stack.push(middleware);
+        self
+    }
+
+    /// Returns a [`Client`] using this builder configuration.
+    pub fn build(self) -> Client {
+        Client::new(self.compilers, self.middleware_stack)
+    }
+}
+
+pub struct Client {
+    compilers: Arc<Compilers<SolidityCompiler>>,
+    middleware_stack: Box<[Arc<dyn Middleware<Success>>]>,
```

Comment:
also why is it `Box<[]>` and not just a vec?

Reply:
In general, it has two advantages: 1) it shrinks the vector to occupy only the space that it fits in (https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice) and does not store the capacity value, thus, requiring less space; 2) it indicates to users that this value is fixed and could not be modified. 

In general, both of these points are not too significant by themselves, but considering that using `Box<[]>` has no overhead, I cannot see a reason not to use that

Reply:
that's just preemptive optimization and imo not needed here. you can easily shrink_to_fit the vec, and we'll save 16 bytes on pointers, which is nothing. 
if we're talking about MiCrOoPtImIzAtIoN, we can talk about shrink_to_fit actually causing the reallocation, so creating the `Box<[]>` from vec kinda has _some_ overhead, and etc etc, but we don't want to talk about that :) 
it seems that the only way to make [boxed slice is from vec](https://stackoverflow.com/a/33704684), I really don't see any reason why not just use vec. It is a standard solution 
but choose whatever you think is better